### PR TITLE
Create an async module that can automatically batch events and dispatch them later.

### PR DESCRIPTION
**Pull request by justinbmeyer**
For #11

### CODE REVIEW

**Comment by matthewp** [comment]
Using setTimeout(fn, 0) imposes a minimum 4ms delay, why wouldn't you use requestAnimationFrame here?

**Comment by justinbmeyer** [comment]
RAF could be larger than 4ms and is about timing paintings.  RAF should be used by can-view-live, but not this.  I'm using setImmediate which we could eventually make use Promises or some other mechanism that actually does `setImmediate`.  But `setImmediate` is the right idea for what this should be doing.

File: async/async.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	},
+	queue: function(){
+		if(!canBatch.collecting()) {
+			canBatch.start();
+			timeout = setImmediate(canBatch.stop);
+		}
+		return syncBatchQueue.apply(this, arguments);
+	},
+	addEventListener: function(){
+		asyncMethods.queue([syncAddEventListener, this, arguments]);
```
Comment (matthewp) [inline async/async.js]: What's the reason to add the event listener async?

Comment (justinbmeyer) [inline async/async.js]: you could get events dispatched before the binding.

```
obj.dispatch("foo");
obj.on("foo", function(){
  // would be called by the above dispatch
})
```

**Comment by matthewp** [comment]
I see, but how will can-view-live subscribe to the faster events?  It subscribes to compute change events, which all goes through this async code, right?

**Comment by justinbmeyer** [comment]
@matthewp I don't know what you mean by "faster events"?

The idea is that can-view-live can get the value of a compute right away, show the value, but the actual binding will happen sometime later.

**Comment by matthewp** [comment]
So we're on the same page here, I'm assuming that with this code:

```js
var mc = compute(2);

mc.on("change", function(){
  // this fired after 4ms
});

mc(3);
```

The change event will fire after 4ms with the current implementation. 

This means when can-view-live listens for a computes change event, as it does here: https://github.com/canjs/can-view-live/blob/master/lib/attr.js#L24

The DOM will only be updated ~4ms after the compute was actually changed.  What you want is for it to render on the next animation frame.

**Comment by justinbmeyer** [comment]
@matthewp yes, that's why live will have its own scheduler based on RAF.  I mentioned this in under the `Alternatives` section.

Having data change on `setImmediate` and DOM change on `RAF` is ideal.  This way data can change often and as soon as possible, ideally completing in-between any animation frames.  The work done by `RAF` should be minimal.  This way the final state can be ready, no computes need to be re-evaluated.  Instead, the DOM updates are simply applied.

**Comment by justinbmeyer** [comment]
You might no be wondering what happens if we use the current `setImmediate` implementation and it schedules events after the `RAF`.

There's a few possibilities:

1.  We get a better `setImmediate` where this is not possible.
2. We have `live`'s `RAF` handler call `async.flush()`
3. We throw our hands up and say ... you DOM changes should have been a bit faster.

**Comment by matthewp** [comment]
I'm still not following.  How can can-view-live's scheduler make compute's "change" event fire in a requestAnimationFrame? 

Here's how it looks like it will work now:

```
(event) -- macrotask --> can-view-live updateDOM()
```

My interpretation of what you are saying is that it will be something like:

```
(event) -- macrotask --> can-view-live -- rAF --> updateDOM()
```

So we're waiting 1 macrotask and then a rAF before a DOM update occurs.

What you actually want is:

```
(event) -- rAF --> can-view-live updateDOM()
```

From the time the compute changes until the time the DOM is updated is 1 requestAnimationFrame (best case scenario, I realize there are reasons why it might be a bit slower if there's a large queue of dom changes for example).

**Pull request merged**